Meter.c

```diff
@@ -276,43 +333,915 @@ static void BarMeterMode_draw(Meter* this, int x, int y, int w) {
-   /*20*/":", /*21*/":", /*22*/":"
-};
+   data->buffer = xReallocArray(data->buffer, nValues, valueSize);
+
+   // Move existing records ("values") to correct position
+   assert(nValues >= data->nValues);
+   size_t moveOffset = (nValues - data->nValues) * nCellsPerValue;
+   GraphDataCell* dest = (GraphDataCell*)data->buffer + moveOffset;
+   memmove(dest, data->buffer, data->nValues * valueSize);
+
+   // Fill new spaces with blank records
+   memset(data->buffer, 0, moveOffset * sizeof(GraphDataCell));
+
+bufferInitialized:
+   data->nValues = nValues;
+}
+
+static size_t GraphMeterMode_valueCellIndex(unsigned int graphHeight, bool isPercentChart, int deltaExp, unsigned int y, unsigned int* scaleFactor, unsigned int* increment) {
+   assert(deltaExp >= 0);
+   assert(deltaExp < UINT16_WIDTH);
+
+   if (scaleFactor)
+      *scaleFactor = 1;
+   if (increment)
+      *increment = isPercentChart ? 1 : (2U << deltaExp);
+
+   unsigned int yTop = (graphHeight - 1) >> deltaExp;
+   if (y > yTop) {
+      assert(increment == NULL);
+      return (size_t)-1;
+   }
+
+   if (isPercentChart) {
+      assert(deltaExp == 0);
+      return y;
+   }
+   // A record may be rendered in different scales depending on the largest
+   // "scaleExp" value of a record set. The colors are precomputed for
+   // different scales of the same record. It takes (2 * graphHeight - 1) cells
+   // of space to store all the color information.
+   //
+   // An example for graphHeight = 6:
+   //
+   //    scale  1*n  2*n  4*n  8*n 16*n | n = value sum of all items
+   // --------------------------------- |     rounded up to a power of
+   // deltaExp    0    1    2    3    4 |     two. The exponent of n is
+   // --------------------------------- |     stored in index [0].
+   //    array [11]    X    X    X    X | X = empty cell
+   //  indices  [9]    X    X    X    X | Cells whose array indices
+   //           [7]    X    X    X    X | are >= (2 * graphHeight) are
+   //           [5] [10]    X    X    X | computed from cells of a
+   //           [3]  [6] (12)    X    X | lower scale and not stored in
+   //           [1]  [2]  [4]  [8] (16) | the array.
+
+   // "b" is the "base" offset or the upper bits of offset
+   unsigned int b = (y * 2) << deltaExp;
+   unsigned int offset = 1U << deltaExp;
+
+   if (!scaleFactor) {
+      // This function is called for writing. The "increment" argument is
+      // optional, but the caller should assert the index is in bounds.
+      if (b + offset > 2 * graphHeight - 1) {
+         assert(increment == NULL);
+         return (size_t)-1;
+      }
+      return b + offset;
+   }
+
+   // This function is called for reading.
+   assert(increment == NULL);
+
+   if (y < yTop)
+      return b + offset;
+
+   assert(((2 * graphHeight - 1) & b) == b);
+   unsigned int offsetTop = powerOf2Floor(2 * graphHeight - 1 - b);
+   if (offsetTop) {
+      *scaleFactor = offset / offsetTop;
+   }
+   return b + offsetTop;
```

Comment:
A bit more space for the code …
```suggestion
   if (y < yTop) {
      return b + offset;
   }

   assert(((2 * graphHeight - 1) & b) == b);

   unsigned int offsetTop = powerOf2Floor(2 * graphHeight - 1 - b);
   if (offsetTop) {
      *scaleFactor = offset / offsetTop;
   }

   return b + offsetTop;
```